design/open/template-server/template-server.md

```diff
@@ -0,0 +1,130 @@
+# Template Server
+
+## About
+
+The template server adds a simple generic templating capability to the
+lakeFS web server.  It adds server-side support for administrators and
+developers to serve templated web objects securely.
+
+Among the planned applications:
+
+* Serve capabilities to clients.  Serving capabilities from a template
+  lets developers add them easily but allows administrators to control
+  what is actually served.
+* Serve prepopulated config files for popular setups.  For instance we
+  can serve `.lakectl` and Hadoop/Spark config files.  These sometimes
+  need to be modified, for instance when lakeFS does not know the name
+  of its endpoint.
+
+  This can be an important part of for time-to-value.  Using templates
+  we offer admins some flexibility in configuration: for instance they
+  might define templates with different values, and clients can select
+  which to receive by path.
+  
+## Template format
+
+We use Golang [text/template][text/template] to expand most templates.
+For browsers (based on filename, e.g. `*.html`) we instead must use
+[html/template][html/template] for safety.
+
+Accesses to http://<lakefs>/api/v1/templates/some/path read the template
+from `some/path`.
+
+The template is read by lakeFS from a specific (configured) prefix, by
+default `lakefs://lakefs-templates/`.  We might decide to support
+various object storage types; `lakefs` gives the best management and lineage
+capabilities, but possible supporting the production blockstore types `s3`,
+`azure` and `gcs` will simplify deployment of some lakeFS deployments.
+
+If storing on lakeFS, IAM authorizes users to see this template by
+`fs:ReadObject`, at which point expansion can start.  If we want to store
+templates on the blockstore types, we will probably need to add a new IAM
+action type for template expansion, or alternatively overload
+`fs:ReadObject` to support off-lakeFS URIs.  We will supply the following
+replacements (at least):
+
+| template function | value source                        | IAM permissions                              |
+|:------------------|:------------------------------------|:---------------------------------------------|
+| config            | exportable lakeFS configuration[^1] | `arn:lakefs:template:::config/path/to/field` |
+| object            | (small) object on lakeFS            | IAM read permission for that object          |
+| contenttype       | none; set `Content-Type:`           | (none)                                       |
+| new_credentials   | new credentials added to user       | `auth:CreateCredentials`                     |
```

Comment:
This would potentially generate multiple credentials for each user. These credentials should be managed in some way.